core/graph/src/main/java/uk/gov/gchq/gaffer/graph/hook/NamedOperationResolver.java

```diff
@@ -122,42 +98,67 @@ public <T> T onFailure(final T result, final OperationChain<?> opChain, final Co
+     * actually exist in the cache and contain their correct {@link Operation}s.
+     * Will also run recursively to a given depth limit to ensure any nested
+     * {@link NamedOperation}s are also resolved from the cache.
+     *
+     * @param operation {@link NamedOperation} or {@link OperationChain} to act on.
+     * @param user      User for the cache access.
+     * @param depth     Current recursive depth, will use limit set in class to
+     *                  continue or not.
+     * @return A list of resolved operations essentially flattened into plain
+     *         Operations and OperationChains.
+     */
+    private Collection<Operation> resolveNamedOperations(final Operation operation, final User user, final int depth) {
+        final Collection<Operation> updatedOperations = new ArrayList<>();
+        LOGGER.debug("Current resolver depth is: {}", depth);
+
+        // If a named operation resolve the operations within it
+        if (operation instanceof NamedOperation) {
+            NamedOperation<?, ?> namedOperation = (NamedOperation<?, ?>) operation;
+            LOGGER.debug("Resolving named operation called: {}", namedOperation.getOperationName());
+            try {
+                // Get the chain for the named operation from the cache
+                final OperationChain<?> namedOperationChain = cache
+                        .getNamedOperation(namedOperation.getOperationName(), user)
+                        .getOperationChain(namedOperation.getParameters());
+                // Update the operation inputs and add operation chain to the updated list
+                OperationHandlerUtil.updateOperationInput(namedOperationChain, namedOperation.getInput());
+
+                // Run again to resolve any nested operations in the chain before adding
+                namedOperationChain.updateOperations(resolveNamedOperations(namedOperationChain, user, depth + 1));
+                updatedOperations.add(namedOperationChain);
+
+            } catch (final CacheOperationException e) {
+                LOGGER.error("Exception resolving NamedOperation within the cache: {}", e.getMessage());
+                return Collections.singletonList(namedOperation);
+            }
+        // If given an operationChain resolve Operations inside it
+        } else if (operation instanceof OperationChain) {
+            LOGGER.debug("Resolving Operations in Operation Chain: {}", ((OperationChain<?>) operation).getOperations());
+            for (final Operation op : ((OperationChain<?>) operation).getOperations()) {
+                // Resolve if haven't hit the depth limit for named operations yet
+                // Note only need to check the depth here as when checking for a nested named operation it will always be
+                // recursively passed as an operation chain
+                if (depth < depthLimit) {
+                    updatedOperations.addAll(resolveNamedOperations(op, user, depth));
```

Comment:
so:

```
OpChain[OpChain[OpChain[OpChain[OpChain[OpChain[NamedOp]]]]]]
```
will still be at the same depth because the depth/nesting of opchains is irrelevant, correct?

Reply:
That's correct it is only the depth of the Named Operation nesting that is of concern